### PR TITLE
Feed response diagnostics in json format

### PR DESCRIPTION
**Pull request by balalakshmin**
# Description
Feed response diagnostics including query metrics used to appear in a text + tabular format which was difficult to parse. Hence changing it to a json format on par with other diagnostics elements.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

Before this change:

image
image

After the change:

Query item diagnostics through iterableByPage : FeedResponseDiagnostics({"queryMetricsMap":{"{\"min\":\"\",\"max\":\"FF\"},pkrId:0":{"retrievedDocumentCount":1,"retrievedDocumentSize":387,"outputDocumentCount":1,"outputDocumentSize":436,"indexHitDocumentCount":1,"totalQueryExecutionTime":"PT0.00047S","queryPreparationTimes":{"queryCompilationTime":"PT0.00012S","logicalPlanBuildTime":"PT0.00003S","physicalPlanBuildTime":"PT0.00006S","queryOptimizationTime":"PT0S"},"indexLookupTime":"PT0.00009S","documentLoadTime":"PT0.00002S","runtimeExecutionTimes":{"queryEngineExecutionTime":"PT0.00003S","systemFunctionExecutionTime":"PT0S","userDefinedFunctionExecutionTime":"PT0S"},"documentWriteTime":"PT0S","clientSideMetrics":{"retries":0,"requestCharge":2.83,"fetchExecutionRanges":[{"startTime":"2023-07-06T14:49:32.703924400Z","endTime":"2023-07-06T14:49:32.714924400Z","partitionId":"{\"min\":\"\",\"max\":\"FF\"}","numberOfDocuments":1,"retryCount":0,"activityId":"d096290a-64af-4c75-805a-801e1f467c5b"}],"partitionSchedulingTimeSpans":[{"{\"min\":\"\",\"max\":\"FF\"}":{"turnaroundTime":21,"responseTime":8,"runTime":11,"waitTime":10,"numPreemptions":1}}]},"indexUtilizationInfo":null,"retries":0,"vmexecutionTime":"PT0.00014S","indexHitRatio":1.0}},"clientSideRequestStatistics":[{"userAgent":"azsdk-java-cosmos/4.47.0-beta.1 Windows11/10.0 JRE/17.0.3","activityId":"d096290a-64af-4c75-805a-801e1f467c5b","requestLatencyInMs":9,"requestStartTimeUTC":"2023-07-06T14:49:32.704672800Z","requestEndTimeUTC":"2023-07-06T14:49:32.713809300Z","responseStatisticsList":[{"storeResult":{"storePhysicalAddress":"rntbd://127.0.0.1:10253/apps/DocDbApp/services/DocDbServer12/partitions/a4cb4958-38c8-11e6-8106-8cdcd42c33be/replicas/1p/","lsn":37,"globalCommittedLsn":-1,"partitionKeyRangeId":"0","isValid":true,"statusCode":200,"subStatusCode":0,"isGone":false,"isNotFound":false,"isInvalidPartition":false,"isThroughputControlRequestRateTooLarge":false,"requestCharge":2.83,"itemLSN":-1,"sessionToken":"0:-1#37","backendLatencyInMs":0.799,"retryAfterInMs":null,"replicaStatusList":["10253:Connected","10253:Unknown","10253:Unknown","10253:Unknown","10253:Unknown","10253:Unknown","10253:Unknown","10253:Connected","10253:Unknown","10253:Unknown","10253:Unknown","10253:Connected"],"transportRequestTimeline":[{"eventName":"created","startTimeUTC":"2023-07-06T14:49:32.709751700Z","durationInMilliSecs":0.0},{"eventName":"queued","startTimeUTC":"2023-07-06T14:49:32.709751700Z","durationInMilliSecs":0.0},{"eventName":"channelAcquisitionStarted","startTimeUTC":"2023-07-06T14:49:32.709751700Z","durationInMilliSecs":0.0},{"eventName":"pipelined","startTimeUTC":"2023-07-06T14:49:32.709751700Z","durationInMilliSecs":1.0098},{"eventName":"transitTime","startTimeUTC":"2023-07-06T14:49:32.710761500Z","durationInMilliSecs":2.0},{"eventName":"decodeTime","startTimeUTC":"2023-07-06T14:49:32.712761500Z","durationInMilliSecs":1.0478},{"eventName":"received","startTimeUTC":"2023-07-06T14:49:32.713809300Z","durationInMilliSecs":0.0},{"eventName":"completed","startTimeUTC":"2023-07-06T14:49:32.713809300Z","durationInMilliSecs":0.0}],"rntbdRequestLengthInBytes":445,"rntbdResponseLengthInBytes":1387,"requestPayloadLengthInBytes":60,"responsePayloadLengthInBytes":436,"channelStatistics":{"channelId":"b6e54e1f","channelTaskQueueSize":1,"pendingRequestsCount":0,"lastReadTime":"2023-07-06T14:49:32.632959400Z","waitForConnectionInit":false},"serviceEndpointStatistics":{"availableChannels":1,"acquiredChannels":0,"executorTaskQueueSize":0,"inflightRequests":1,"lastSuccessfulRequestTime":"2023-07-06T14:49:32.633Z","lastRequestTime":"2023-07-06T14:49:32.629Z","createdTime":"2023-07-06T14:49:32.471982700Z","isClosed":false,"cerMetrics":{}}},"requestResponseTimeUTC":"2023-07-06T14:49:32.713809300Z","requestStartTimeUTC":"2023-07-06T14:49:32.704672800Z","requestResourceType":"Document","requestOperationType":"Query","requestSessionToken":null}],"supplementalResponseStatisticsList":[],"addressResolutionStatistics":{},"regionsContacted":["south central us"],"retryContext":{"statusAndSubStatusCodes":null,"retryCount":0,"retryLatency":0},"metadataDiagnosticsContext":{"metadataDiagnosticList":null},"serializationDiagnosticsContext":{"serializationDiagnosticsList":null},"gatewayStatistics":null,"samplingRateSnapshot":1.0,"systemInformation":{"usedMemory":"17104 KB","availableMemory":"8301872 KB","systemCpuLoad":"empty","availableProcessors":8},"clientCfgs":{"id":1,"machineId":"uuid:46cd26f5-4112-4c48-8754-e06f9be426e9","connectionMode":"DIRECT","numberOfClients":1,"clientEndpoints":{"https://localhost:8081":1},"connCfg":{"rntbd":"(cto:PT5S, nrto:PT5S, icto:PT0S, ieto:PT1H, mcpe:130, mrpc:30, cer:true)","gw":"(cps:1000, nrto:PT1M, icto:PT1M, p:false)","other":"(ed: true, cs: false, rv: true)"},"consistencyCfg":"(consistency: Eventual, mm: true, prgns: [])","proactiveInit":null}}],"userAgent":"azsdk-java-cosmos/4.47.0-beta.1 Windows11/10.0 JRE/17.0.3","queryPlanDiagnosticsContext":{"startTimeUTC":"2023-07-06T14:49:32.664002700Z","endTimeUTC":"2023-07-06T14:49:32.675531900Z","requestTimeline":[{"eventName":"connectionAcquired","startTimeUTC":"2023-07-06T14:49:32.669064Z","durationInMilliSecs":1.2053},{"eventName":"connectionConfigured","startTimeUTC":"2023-07-06T14:49:32.670269300Z","durationInMilliSecs":0.0},{"eventName":"requestSent","startTimeUTC":"2023-07-06T14:49:32.670269300Z","durationInMilliSecs":1.0157},{"eventName":"transitTime","startTimeUTC":"2023-07-06T14:49:32.671285Z","durationInMilliSecs":2.2497},{"eventName":"received","startTimeUTC":"2023-07-06T14:49:32.673534700Z","durationInMilliSecs":1.0002}]}})

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedResponseDiagnostics.java [hunk 1]
```diff
@@ -66,49 +67,7 @@ public Map<String, QueryMetrics> getQueryMetricsMap() {
...
-        }
-        try {
-            stringBuilder
-                .append(mapper.writeValueAsString(clientSideRequestStatistics));
-        } catch (JsonProcessingException e) {
-            LOGGER.error("Error while parsing diagnostics ", e);
-        }
-
-        return stringBuilder.toString();
+        return RntbdObjectMapper.toString(this);
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedResponseDiagnostics.java:70]: why are we using RntbdObjectMapper? We should re-use the one which is already defined in this class.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/QueryMetrics.java [hunk 2]
```diff
@@ -295,6 +296,6 @@ public static QueryMetrics createFromDelimitedStringAndClientSideMetrics(String
 
     @Override
     public String toString() {
-        return toTextString(0);
+        return RntbdObjectMapper.toString(this);
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/QueryMetrics.java:299]: Same here, if there is no objectMapper, we should use the `Utils.getObjectMapper()`

**Review by kushagraThapar — CHANGES_REQUESTED** [review]
The changes look good, however, we used to have so many classes for converting query metrics to tabular format. Should we keep those or delete the classes that are not being used anymore in the Java SDK?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedResponseDiagnostics.java (hunk 1 above)
Comment (balalakshmin) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedResponseDiagnostics.java:70]: RntbdObjectMapper also uses the object mapper only internally. But it abstracts a few exception handling. And I saw this being used in a similar case. But if it has to be used only within Rtnbd packages, I can change it.

Comment (balalakshmin) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedResponseDiagnostics.java:70]: It also has additional configuration for object mapper without which using the plain objectMapper fails serialization. 
(https://github.com/Azure/azure-sdk-for-java/blob/b3679847c498627c8a7900e9f5d66e562e9f2c5a/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdObjectMapper.java#L38C5-L43C44)

Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedResponseDiagnostics.java:64]: Got it, makes sense. 
@xinlian12  can you pls confirm if it is safe to use RntbdObjectMapper outside of rntbd package?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/QueryMetrics.java (hunk 2 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/QueryMetrics.java:290]: does this also fails on serialization if we try to use the `Utils.getObjectMapper()`?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedResponseDiagnostics.java (hunk 1 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedResponseDiagnostics.java:64]: I still think we should not use it outside of the rntbd package. I see we have one in this class, can't we customize it accordingly and use it here locally?

**Review by kushagraThapar — CHANGES_REQUESTED** [review]

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedResponseDiagnostics.java (hunk 1 above)
Comment (balalakshmin) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedResponseDiagnostics.java:64]: Yes, we can do that.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/QueryMetrics.java (hunk 2 above)
Comment (balalakshmin) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/QueryMetrics.java]: I've changed the objectMapper config in FeedResponse to handle Duration types.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/QueryMetrics.java [hunk 3]
```diff
@@ -295,6 +288,10 @@ public static QueryMetrics createFromDelimitedStringAndClientSideMetrics(String
 
     @Override
     public String toString() {
-        return toTextString(0);
+        try {
+            return Utils.getSimpleObjectMapper().writeValueAsString(this);
+        } catch (final JsonProcessingException error) {
+            return "null";
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/QueryMetrics.java:294]: Should we log the error?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedResponseDiagnostics.java [hunk 4]
```diff
@@ -4,26 +4,30 @@
...
+
 /**
  * The type Feed response diagnostics.
  */
 public class FeedResponseDiagnostics {
 
-    private final static String EQUALS = "=";
-    private final static String QUERY_PLAN = "QueryPlan";
-    private final static String SPACE = " ";
     private static final ObjectMapper mapper = new ObjectMapper();
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedResponseDiagnostics.java:29]: On second thought, I am curious if we can re-use the `Utils.getSimpleObjectMapper()`?

**Review by kushagraThapar — COMMENTED** [review]
Looks good @balalakshmin , keeping it in request changes mode for now until we release cosmos java SDK this week.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedResponseDiagnostics.java (hunk 4 above)
Comment (balalakshmin) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedResponseDiagnostics.java:29]: Utils.getSimpleObjectMapper might be used in multiple places. And we have added additional serialization and deserialization config. Hence preferred to keep it custom within this class.

Comment (balalakshmin) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedResponseDiagnostics.java]: I've added another helper method in the Utils to create an object mapper with duration configs enabled and using that now

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Utils.java [hunk 5]
```diff
@@ -101,6 +105,17 @@ private static ObjectMapper createAndInitializeObjectMapper(boolean allowDuplica
...
     }
 
+    private static ObjectMapper createAndInitializeDurationObjectMapper() {
+        ObjectMapper objectMapper = new ObjectMapper();
+        SimpleFilterProvider filterProvider = new SimpleFilterProvider();
+        objectMapper.registerModule(new SimpleModule()
+                .addSerializer(Duration.class, ToStringSerializer.instance)
+                .addDeserializer(Duration.class, DurationDeserializer.INSTANCE)
+                .addSerializer(Instant.class, ToStringSerializer.instance))
+            .setFilterProvider(filterProvider);
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Utils.java:115]: why do we need the filter provider?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Utils.java [hunk 6]
```diff
@@ -368,6 +383,10 @@ public static ObjectMapper getSimpleObjectMapper() {
         return Utils.simpleObjectMapper;
     }
 
+    public static ObjectMapper getDurationEnabledObjectMapper() {
+        return Utils.durationEnabledObjectMapper;
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Utils.java:387]:
```suggestion
        return durationEnabledObjectMapper;
```

**Comment by kushagraThapar** [comment]
@balalakshmin - seems like the json object is broken (from semantic point of view) - can you please verify, thanks!

**Review by kushagraThapar — APPROVED** [review]

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedResponseDiagnostics.java [hunk 7]
```diff
@@ -4,26 +4,27 @@
 package com.azure.cosmos.implementation;
 
 import com.azure.cosmos.implementation.query.QueryInfo;
-import com.azure.cosmos.implementation.query.metrics.QueryMetricsTextWriter;
 import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.module.SimpleModule;
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedResponseDiagnostics.java:9]: Unused imports?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Utils.java (hunk 5 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Utils.java:115]: Just curious - the deserializer for `Duration` - is this needed for a `objectMapper` primarily used in `toString()`? If so, why no deserializer for `Instant`?

File: sdk/cosmos/azure-cosmos/CHANGELOG.md [hunk 8]
```diff
@@ -24,6 +24,7 @@
   with the region in the direct connectivity mode. - See [PR 35586](https://github.com/Azure/azure-sdk-for-java/pull/35586)
 
 #### Other Changes
+* Query metrics diagnostics changed to JSON format. - See [PR 35761](https://github.com/Azure/azure-sdk-for-java/pull/35761)
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos/CHANGELOG.md:27]: Can be moved to the `4.49.0` - let me confirm this with @kushagraThapar

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/CosmosDiagnosticsTest.java [hunk 9]
```diff
@@ -462,19 +462,16 @@ public void queryPlanDiagnostics() throws JsonProcessingException {
                 FeedResponse<InternalObjectNode> feedResponse = iterator.next();
                 queryDiagnostics = feedResponse.getCosmosDiagnostics().toString();
                 if (feedResponseCounter == 0) {
-                    assertThat(queryDiagnostics).contains("QueryPlan Start Time (UTC)=");
-                    assertThat(queryDiagnostics).contains("QueryPlan End Time (UTC)=");
-                    assertThat(queryDiagnostics).contains("QueryPlan Duration (ms)=");
+                    assertThat(queryDiagnostics).contains("queryPlanDiagnosticsContext");
+                    assertThat(queryDiagnostics).contains("startTimeUTC");
+                    assertThat(queryDiagnostics).contains("endTimeUTC");
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/CosmosDiagnosticsTest.java:467]: I am curious if we have an out of the box way to compute duration here.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Utils.java (hunk 5 above)
Comment (balalakshmin) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Utils.java:115]: These are the configs used in RntbdObjectMapper. Serialization works as expected without filter provider and deserializer settings as well. So I've removed them.

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/CosmosDiagnosticsTest.java (hunk 9 above)
Comment (balalakshmin) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/CosmosDiagnosticsTest.java:467]: Yes we do have a durationInMilliSecs field. Added that as well here for verification

**Review by jeet1995 — APPROVED** [review]
LGTM - thanks for the change!

**Pull request merged**